### PR TITLE
Construct URLSearchParams from array or object

### LINKED ISSUE #22554: Construct URLSearchParams from array or object
whatwg/url#175
web-platform-tests/wpt#4523

Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1330678

### LINKED ISSUE #22556: Intermittent FAIL in /url/urlsearchparams-constructor.any.html, /url/urlsearchparams-constructor.any.worker.html
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_array_equals: property 0, expected \"c\" but got \"a\"", 
    "stack": "@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:170:13\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1566:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:166:5\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:160:1\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.worker.js:8:1\n", 
    "subtest": "Construct with object with two keys", 
    "test": "/url/urlsearchparams-constructor.any.worker.html", 
    "line": 128495, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_array_equals: property 0, expected \"a\\0b\" but got \"d\u1234\"", 
    "stack": "@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:170:13\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1566:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:166:5\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:160:1\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.worker.js:8:1\n", 
    "subtest": "Construct with object with NULL, non-ASCII, and surrogate keys", 
    "test": "/url/urlsearchparams-constructor.any.worker.html", 
    "line": 128497, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_array_equals: property 0, expected \"a\\0b\" but got \"d\u1234\"", 
    "stack": "@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:170:13\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1566:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:166:5\n@http://web-platform.test:8000/url/urlsearchparams-constructor.any.js:160:1\n", 
    "subtest": "Construct with object with NULL, non-ASCII, and surrogate keys", 
    "test": "/url/urlsearchparams-constructor.any.html", 
    "line": 185202, 
    "action": "test_result", 
    "expected": "PASS"
}
```

### LINKED ISSUE #22557: Use IndexMap for WebIDL record type
Currently, when we codegen the `record` type of `WebIDL`, we'll use `MozMap` for it.

https://github.com/servo/servo/blob/47a308b277f1023f5fda0c3db465317a8691818a/components/script/dom/bindings/codegen/CodegenRust.py#L2036-L2037

While implementing #22555 , I found there will be two cases `FAIL`ed intermittently (#22556).

So, the root cause is, per the spec of [WebIDL `record` type](https://heycam.github.io/webidl/#idl-record), it should be `ordered maps`. However, we're currently using `HashMap` which is an unordered collection.

https://github.com/servo/servo/blob/47a308b277f1023f5fda0c3db465317a8691818a/components/script/dom/bindings/mozmap.rs#L28-L32

~We should use `BTreeMap` which is an ordered collection instead.~
(We cannot use `BTreeMap` because it will be sorted by keys while iterating.)
Ref: https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.iter

We should use [`IndexMap` from `indexmap` crate](https://docs.rs/indexmap/1.0.2/indexmap/map/struct.IndexMap.html) which is an ordered collection and it will also preserve the order while iterating.

Once this is done, #22556 should be fixed.

### PR DESCRIPTION
**Pull request by CYBAI**
---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22554, fix #22556 and also fix #22557 
- [x] There are tests in `url/urlsearchparams-constructor.any.js` for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22555)

### CODE REVIEW

File: tests/wpt/metadata/url/urlsearchparams-constructor.any.js.ini [hunk 1]
```diff
@@ -1,51 +1,8 @@
 [urlsearchparams-constructor.any.worker.html]
-  [Construct with object with two keys]
-    expected: FAIL
-
-  [Construct with object with NULL, non-ASCII, and surrogate keys]
```
Comment (CYBAI) [inline tests/wpt/metadata/url/urlsearchparams-constructor.any.js.ini]: I've observed the `Construct with object with two keys` case and `Construct with object with NULL, non-ASCII, and surrogate keys` case will be `PASS`ed intermittently in my local.

I belive the root cause would be related to the `MozMap` type is `HashMap` which is an unordered collection so we might get result in different order when using map 🤔

Comment (jdm) [reply tests/wpt/metadata/url/urlsearchparams-constructor.any.js.ini]: That sounds like a really good idea!

**Comment by jdm** [comment]
That failure is #13480.

File: components/script/dom/urlsearchparams.rs [hunk 2]
```diff
@@ -47,13 +47,30 @@ impl URLSearchParams {
...
     ) -> Fallible<DomRoot<URLSearchParams>> {
         // Step 1.
         let query = URLSearchParams::new(global, None);
         match init {
-            Some(USVStringOrURLSearchParams::USVString(init)) => {
+            Some(USVStringSequenceSequenceOrUSVStringUSVStringRecordOrUSVString::USVStringSequenceSequence(init)) => {
                 // Step 2.
+
+                // Step 2-1.
+                if init.iter().filter(|pair| pair.len() == 2).count() != init.len() {
```
Comment (jdm) [inline components/script/dom/urlsearchparams.rs]: `if init.iter().any(|pair| pair.len() != 2) {`

**Review by jdm — CHANGES_REQUESTED** [review]

File: tests/wpt/metadata/url/urlsearchparams-constructor.any.js.ini (hunk 1 above)
Comment (CYBAI) [reply tests/wpt/metadata/url/urlsearchparams-constructor.any.js.ini]: Yeah, let me file an issue for it! :)

**Comment by CYBAI** [comment]
@bors-servo retry

-  #22556 (that is the intermittent I mentioned in https://github.com/servo/servo/pull/22555#discussion_r243912934)

File: tests/wpt/metadata/url/urlsearchparams-constructor.any.js.ini (hunk 1 above)
Comment (CYBAI) [reply tests/wpt/metadata/url/urlsearchparams-constructor.any.js.ini]: Filed at #22557

**Comment by CYBAI** [comment]
@bors-servo retry

- #22556

**Comment by jdm** [comment]
It's a bit sad to mark the test as intermittent when this PR is fixing a bunch of its tests. Why not change the MozMap implementation in this PR instead?

**Comment by CYBAI** [comment]
@jdm I wondered if it's not good to change MozMap in this PR because I thought it's not related to URLSearchParams itself. But, I also start to think that it should be fine to fix it in this PR and maybe I can separate them into different commits to say why we need to change the MozMap. So, I'm working on fixing the MozMap! Thanks for telling me that you think it's fine to fix MozMap in this PR :)

**Comment by CYBAI** [comment]
@jdm With testing to use `BTreeMap` locally, I found we can pass `Construct with object with NULL, non-ASCII, and surrogate keys` consistently and fail `Construct with object with two keys` consistently.

The reason why we'll fail `Construct with object with two keys` is, while iterating `BTreeMap`, the map will be [ordered by key](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.iter). For example, if keys are inserted with `c` and then `a`, it will be iterated with `a` and then `c`.

After doing more investigation, maybe what we want for the `orderer map` for `WebIDL record` is something like [indexmap](https://docs.rs/indexmap/1.0.2/indexmap/map/struct.IndexMap.html) ? 

Would you think it's fine to install that crate?

**Comment by CYBAI** [comment]
@bors-servo try=wpt

- Just tried to use `IndexMap` and the tests can `PASS` consistently in my local :)

> It's a bit sad to mark the test as intermittent when this PR is fixing a bunch of its tests. Why not change the MozMap implementation in this PR instead?

Sorry again and yeah, I should tryt to avoid introducing a known intermittent in the future, thanks!

**Comment by CYBAI** [comment]
@bors-servo try=wpt

- Let's try againg to see if we'll get that intermittents 👀

File: components/script/dom/urlsearchparams.rs [hunk 3]
```diff
@@ -66,13 +83,10 @@ impl URLSearchParams {
                 *query.list.borrow_mut() =
                     form_urlencoded::parse(init_bytes).into_owned().collect();
             },
-            Some(USVStringOrURLSearchParams::URLSearchParams(init)) => {
-                // Step 3.
-                *query.list.borrow_mut() = init.list.borrow().clone();
-            },
-            None => {},
+            _ => {},
```
Comment (emilio) [inline components/script/dom/urlsearchparams.rs]: This branch should probably still use `None`, should't it?

Comment (CYBAI) [reply components/script/dom/urlsearchparams.rs]: Ah, yeah! Good catch! I forgot to change it back while testing each implementation, thanks!

**Comment by jdm** [comment]
@bors-servo r+
Thanks for fixing that!

**Pull request merged**